mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassinListener.java

```diff
@@ -147,37 +163,46 @@ private boolean isAppendedToInbox(Added addedEvent) {
         }
     }
 
-    private ImmutableList<InputStream> retrieveMessages(MessageMoveEvent messageMoveEvent, MailboxSession session) {
-        return mapperFactory.getMessageIdMapper(session)
-            .find(messageMoveEvent.getMessageIds(), MessageMapper.FetchType.Full)
-            .stream()
-            .map(Throwing.function(Message::getFullContent))
-            .collect(Guavate.toImmutableList());
+    private Flux<MessageToLearn> retrieveMessages(MessageMoveEvent messageMoveEvent, MailboxSession session) {
+        return Flux.defer(() ->
+            Flux.fromIterable(mapperFactory.getMessageIdMapper(session)
+                .find(messageMoveEvent.getMessageIds(), MessageMapper.FetchType.Full)))
+            .map((ThrowingFunction<MailboxMessage, MessageToLearn>) message -> new MessageToLearn(message.getFullContent(), message.getFullContentOctets()))
+            .subscribeOn(Schedulers.elastic());
     }
 
     @VisibleForTesting
-    boolean isMessageMovedToSpamMailbox(MessageMoveEvent event) {
-        try {
-            MailboxId spamMailboxId = systemMailboxesProvider.findMailbox(Role.SPAM, event.getUsername()).getId();
-
-            return event.getMessageMoves().addedMailboxIds().contains(spamMailboxId);
-        } catch (MailboxException e) {
-            LOGGER.warn("Could not resolve Spam mailbox", e);
-            return false;
-        }
+    Mono<Boolean> isMessageMovedToSpamMailbox(MessageMoveEvent event) {
+        return findMailboxByRole(Role.SPAM, event.getUsername())
```

Comment:
it should be done at an higher level because we look for findMailboxByRole many times for the same mailbox

Reply:
I didn't change the behavior of that. But i don't see how you would do that.
Here we do it only once per event and not once per message. Maybe the name of the method is not good and it should be `areMessagesMovedToSpamMailbox` ?

Reply:
oh, ok, if it's done only once, naming is misleading.
What about returning a Flux of messages that are moved to spam? I'm pretty sure it would make everything nicer

Reply:
It is an interesting idea. But i think it is another refactor out of the scope of this PR.